test_plans/SYCL-2020/aspect_selector.md

```diff
@@ -0,0 +1,127 @@
+
+Additional testing will include 100 combinations
+of randomly selected aspect combinations,
+allowing aspect duplicates,
+with a random arity.
+The random combinations are selected at test generation time,
+with each test generation producing the same combinations
+and each run testing the same random combinations.
+
+For each of the combinations described so far,
+add new combinations which also use forbidden aspects `{noasp1, ..., noaspM}` -
+`M` is independent of `N`, can be smaller, same size, or larger.
+These forbidden aspects will be passed to the `denyList` argument.
+Use the same rules as above for generating aspects,
+following these extra restrictions:
+
+* `noaspY` is not a member of the set `{asp1, ..., aspN}`
+    for each `noaspY` from the set of forbidden aspects.
+
+> Note: it is recommended a Python generator is written to cover all required combinations.
+
+## Tests
+
+### No aspects
+
+1. `auto selector1 = aspect_selector()`
+1. `auto selector2 = aspect_selector({})`
+1. `auto selector3 = aspect_selector({}, {})`
+1. `auto selector4 = aspect_selector<>()`
+
+In the test, all constructed selectors `selectorX` are used
+to construct a SYCL device, `devX` for each selector respectively.
+All devices `devX` must compare equal to the device obtained by `default_selector`
+(let's call this device `defaultDevice`).
+
+### Requested aspects using a device
+
+Assume requested aspects `{asp1, ..., aspN}` for each test,
+and forbidden aspects `denyList = {noasp1, ..., noaspM}`.
+
+These steps are performed before testing combinations:
+
+1. Iterate over all devices to find out if there's one that contains all requested aspects
+    but doesn't contain any forbidden ones,
+    we call this `suitableDevice`.
+
+    * If a device like this cannot be found, skip test.
+
+1. Iterate over all platforms to find out if there's one that contains all requested aspects
+    but doesn't contain any forbidden ones,
+    we call this `suitablePlatform`.
+
+    * If a platform like this cannot be found,
+      continue testing but ignore further tests for the platform.
+
+These steps are performed for each combination:
+
+1. Test using a vector of aspects
+
+    1. Construct `selectorVector` object obtained by calling `aspect_selector(std::vector{asp1, ..., aspN})`.
+    1. Construct SYCL device (`selectedDevice`) using `selectorVector`.
+    1. Construct SYCL queue (`selectedQueue`) using `selectorVector`.
+    1. Check that `selectedDevice` contains all requested aspects
+        and none of the forbidden ones.
+
+        * `selectedDevice.has(asp1), ..., selectedDevice.has(aspN)`.
+        * `!selectedDevice.has(noasp1), ..., !selectedDevice.has(noaspN)`.
+
+            * If `denyList` is not empty.
+        * The same checks for `selectedQueue.get_device()`.
+    1. If `suitablePlatform` was found, construct SYCL platform (`selectedPlatform`) using `selectorVector`.
+    1. Check that `selectedPlatform` contains all requested aspects.
```

Comment:
I think this is not correct.  Here is the definition of the platform constructor you are testing:

```
template <typename DeviceSelector>
explicit platform(const DeviceSelector &deviceSelector);
```

> Constructs a SYCL platform instance that is a copy of the platform which **contains the device** returned by the device selector parameter.

(Emphasis mine.)

The returned platform is merely guaranteed to contain **a** device that has all the requested aspects.  However, this does not mean that the platform itself has all of those aspects.  This is because `platform::has(asp)` is only `true` if **all** devices in the platform have aspect `asp`.

Note that this also affects your pre-condition where you check `suitablePlatform`.